libsoftwarecontainer/src/netlink.cpp

```diff
@@ -0,0 +1,573 @@
+        ifaddrmsg ifaddr = addr.first;
+        if (!ifaddr.ifa_index == ifaceIndex) {
+            continue;
+        }
+
+        netlink_request<ifaddrmsg> addr_msg = createMessage<ifaddrmsg>(RTM_DELLINK, 0);
+        memcpy(&addr_msg.pay, &ifaddr, sizeof(ifaddrmsg));
+        sendMessage(addr_msg);
+    }
+
+    // Then, remove the actual links.
+    for (LinkInfo link : m_links) {
+        ifinfomsg ifinfo = link.first;
+        if(ifinfo.ifi_type == ARPHRD_LOOPBACK) {
+            continue; // This is the loopback device
+        }
+
+        netlink_request<ifinfomsg> down_msg = createMessage<ifinfomsg>(RTM_NEWLINK, 0);
+        down_msg.pay.ifi_family = AF_UNSPEC;
+        down_msg.pay.ifi_index = ifinfo.ifi_index;
+        down_msg.pay.ifi_flags = ~IFF_UP;
+        down_msg.pay.ifi_change = IFF_UP;
+        sendMessage(down_msg);
+    }
+    return true;
+
+}
+
+bool Netlink::isBridgeAvailable(const char *bridgeName, const char *expectedAddress)
+{
+    if (!m_hasKernelDump && !getKernelDump()) {
+        fprintf(stderr, "Could not get cache dump from kernel\n");
+        return false;
+    }
+
+    bool hasBridge = false;
+    unsigned int bridge_ifindex = -1;
+
+    for (LinkInfo linkinfo : m_links) {
+        AttributeList attributes = linkinfo.second;
+        for (AttributeInfo attrinfo : attributes) {
+            rtattr attr = attrinfo.first;
+            void *data = attrinfo.second;
+
+            if (attr.rta_type == IFLA_IFNAME) {
+                char *ifname = (char *) data;
+                if (strcmp(ifname, bridgeName) == 0) {
+                    hasBridge = true;
+                    bridge_ifindex = linkinfo.first.ifi_index;
+                    break;
+                }
+            }
+        }
+
+        // TODO: Make this nicer, please
+        if (hasBridge) {
+            break;
+        }
+    }
+
+    if (!hasBridge) {
+        return false;
+    }
+
+    bool bridgeHasGateway = false;
+    for (AddressInfo addressInfo : m_addresses) {
+        ifaddrmsg addrmsg = addressInfo.first;
+        if (addrmsg.ifa_index != bridge_ifindex) {
+            continue;
+        }
+
+        AttributeList attributes = addressInfo.second;
+        for (AttributeInfo attrPair : attributes) {
+            rtattr attr = attrPair.first;
+            if (attr.rta_type != IFA_ADDRESS && attr.rta_type != IFA_LOCAL) {
+                continue;
+            }
+
+            void *data = attrPair.second;
+            char out[INET6_ADDRSTRLEN];
+            // It is ok here for inet_ntop to fail (data could be bad)
+            if (inet_ntop(addrmsg.ifa_family, data, out, sizeof(out))) {
+                if (strcmp(out, expectedAddress) == 0) {
+                    bridgeHasGateway = true;
+                    break;
+                }
+            }
+        }
+
+        if (bridgeHasGateway) {
+            break;
+        }
+    }
+
+    return bridgeHasGateway;
+}
+
+int Netlink::readMessage()
```

Comment:
This method should maybe be split into multiple parts. E.g. One method dealing with link, one method dealing with routes ....  Or maybe it could be solved with some nice template magic ? Thinking mostly of the switch statement.